FileIOUtils.cpp

```diff
@@ -0,0 +1,426 @@
+                                                    QObject::tr("bin file (*.bin)"));
+
+    // load data into QBytearray
+    QByteArray tmptile8x8data, tmptile8x8data_final;
+    QFile gfxbinfile(fileName);
+    if(!gfxbinfile.open(QIODevice::ReadOnly))
+    {
+        QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("Cannot open file! \n").append(gfxbinfile.errorString()));
+        return false;
+    }
+    if(!gfxbinfile.size())
+    {
+        QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("File size is 0!"));
+        return false;
+    }
+    tmptile8x8data = gfxbinfile.readAll();
+    tmptile8x8data_final = tmptile8x8data; // Init
+    gfxbinfile.close();
+
+    // Check size
+    if(tmptile8x8data.size() & 31)
+    {
+        QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("Illegal file size!\nIt should be a multiple of 32 Bytes."));
+        return false;
+    }
+
+    // Load palette data from bin file
+    fileName = QFileDialog::getOpenFileName(parent,
+                                            QObject::tr("Load palette bin file"), QString(""),
+                                            QObject::tr("bin file (*.bin)"));
+    QByteArray tmppalettedata;
+    QFile palbinfile(fileName);
+    if(!palbinfile.open(QIODevice::ReadOnly))
+    {
+        QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("Cannot open file!"));
+        return false;
+    }
+    tmppalettedata = palbinfile.readAll();
+    palbinfile.close();
+
+    QVector<QRgb> tmppalette;
+    unsigned short *tmppaldata = new unsigned short[16];
+    memset(tmppaldata, 0, 32);
+    memcpy(tmppaldata, tmppalettedata.data(), qMin(32, tmppalettedata.size()));
+    ROMUtils::LoadPalette(&tmppalette, tmppaldata, true);
+    delete[] tmppaldata;
+
+    // Get transparent color id in the palette
+    int transparentcolorId = 0;
+    SelectColorDialog scdialog;
+    scdialog.SetPalette(tmppalette);
+    scdialog.SetColor(0);
+    if(scdialog.exec() == QDialog::Accepted)
+    {
+        transparentcolorId = scdialog.GetSelectedColorId();
+    } else {
+        return false;
+    }
+
+    // transparent-substitute color replacement and load palette
+    tmppalette[transparentcolorId] = 0;
+
+    // nybble exchange not needed
+    // reset bytearray according to the palette bin file
+    for(int i = 0; i != 16; ++i)
+    {
+        char count = 0;
+
+        // Find if the color[i] is in the current palette
+        while(1)
+        {
+            if(tmppalette[i] == ref_palette[count])
+            {
+                break;
+            }
+            ++count;
+            if(count == 16)
+            {
+                if((tmppalette[i] != 0xFF000000) && (tmppalette[i] != 0xFFFFFFFF) && (tmppalette[i] != 0))
+                {
+                    QMessageBox::critical(parent, QObject::tr("Error"), QObject::tr("Palette not suitable!"));
```

Comment:
Can you be more descriptive? Why is it not suitable

Reply:
done